File: reactor-blender-bridge/reactor_bridge/src/protocol.rs
// =============================================================================
// Protocol — REACTOR Bridge (PROTOCOL_VERSION = 1)
// =============================================================================
// Mensajes JSON internally-tagged compartidos entre el servidor Rust y los
// clientes (Blender addon Python, tester standalone, futuro web preview).
//
// Spec completa: ../../proto/messages.md
// =============================================================================

use serde::{Deserialize, Serialize};

/// Versión del protocolo. Incrementar en cualquier cambio breaking.
pub const PROTOCOL_VERSION: u32 = 1;

/// Nombre del servidor (enviado en HelloAck).
pub const SERVER_NAME: &str = "reactor_bridge";

/// Capabilities que el servidor de FASE 0 soporta.
pub const SERVER_CAPABILITIES: &[&str] = &["ping"];

// -----------------------------------------------------------------------------
// Sobre principal
// -----------------------------------------------------------------------------

/// Mensaje top-level — serializa como `{"type": "Foo", "data": {...}}`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum Message {
    Hello(Hello),
    HelloAck(HelloAck),
    Ping(Ping),
    Pong(Pong),
    Error(ErrorPayload),
    Goodbye(Goodbye),
}

impl Message {
    /// Serializa a JSON UTF-8 listo para enviar por el WebSocket.
    pub fn to_json(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    /// Parsea desde un string JSON recibido del WebSocket.
    pub fn from_json(s: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(s)
    }
}

// -----------------------------------------------------------------------------
// Hello / HelloAck — handshake
// -----------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hello {
    pub version: u32,
    pub client: String,
    #[serde(default)]
    pub capabilities: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HelloAck {
    pub version: u32,
    pub server: String,
    pub accepted: bool,
    #[serde(default)]
    pub capabilities: Vec<String>,
    #[serde(default)]
    pub reason: Option<String>,
}

// -----------------------------------------------------------------------------
// Ping / Pong — latencia
// -----------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Ping {
    pub seq: u32,
    pub ts_micros: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pong {
    pub seq: u32,
    pub client_ts_micros: i64,
    pub server_ts_micros: i64,
}

// -----------------------------------------------------------------------------
// Error / Goodbye
// -----------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Goodbye {
    #[serde(default)]
    pub reason: String,
}

// -----------------------------------------------------------------------------
// Re-exports cómodos
// -----------------------------------------------------------------------------

pub use ErrorPayload as Error;

// -----------------------------------------------------------------------------
// Códigos de error estándar (ver proto/messages.md)
// -----------------------------------------------------------------------------

pub mod codes {
    pub const INCOMPATIBLE_VERSION: &str = "INCOMPATIBLE_VERSION";
    pub const UNKNOWN_MESSAGE: &str = "UNKNOWN_MESSAGE";
    pub const MALFORMED_PAYLOAD: &str = "MALFORMED_PAYLOAD";
    pub const NOT_HANDSHAKED: &str = "NOT_HANDSHAKED";
    pub const INTERNAL: &str = "INTERNAL";
}

// -----------------------------------------------------------------------------
// Helper de timestamps (microsegundos epoch)
// -----------------------------------------------------------------------------

/// Devuelve el timestamp actual en microsegundos desde UNIX_EPOCH.
pub fn now_micros() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_micros() as i64)
        .unwrap_or(0)
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hello_roundtrip() {
        let m = Message::Hello(Hello {
            version: 1,
            client: "blender_addon".into(),
            capabilities: vec!["scene_push".into()],
        });
        let json = m.to_json().unwrap();
        let back = Message::from_json(&json).unwrap();
        match back {
            Message::Hello(h) => {
                assert_eq!(h.version, 1);
                assert_eq!(h.client, "blender_addon");
                assert_eq!(h.capabilities, vec!["scene_push".to_string()]);
            }
            _ => panic!("expected Hello"),
        }
    }

    #[test]
    fn pong_roundtrip() {
        let m = Message::Pong(Pong {
            seq: 7,
            client_ts_micros: 1000,
            server_ts_micros: 1500,
        });
        let json = m.to_json().unwrap();
        assert!(json.contains("\"type\":\"Pong\""));
        let back = Message::from_json(&json).unwrap();
        match back {
            Message::Pong(p) => {
                assert_eq!(p.seq, 7);
                assert_eq!(p.client_ts_micros, 1000);
                assert_eq!(p.server_ts_micros, 1500);
            }
            _ => panic!("expected Pong"),
        }
    }

    #[test]
    fn unknown_type_fails_cleanly() {
        let json = r#"{"type":"NotARealMessage","data":{}}"#;
        assert!(Message::from_json(json).is_err());
    }

    #[test]
    fn version_constant_is_one() {
        assert_eq!(PROTOCOL_VERSION, 1);
    }
}

File: reactor-blender-bridge/python/transport/protocol.py
"""Protocol definitions for REACTOR Bridge (PROTOCOL_VERSION = 1).

Mirrors the Rust `reactor_bridge::protocol` types for use from
Python clients (Blender addon, test harness, CLI tools).

Spec: ../../proto/messages.md
"""

import json
import time
from dataclasses import dataclass, field, asdict
from typing import Optional

PROTOCOL_VERSION: int = 1

SERVER_NAME: str = "reactor_bridge"
SERVER_CAPABILITIES: list[str] = field(default_factory=lambda: ["ping"])


class MessageType:
    """Canonical message type strings matching Rust serde tags."""

    HELLO = "Hello"
    HELLO_ACK = "HelloAck"
    PING = "Ping"
    PONG = "Pong"
    ERROR = "Error"
    GOODBYE = "Goodbye"
    TRANSFORM_UPDATED = "TransformUpdated"
    ENTITY_CREATED = "EntityCreated"
    ENTITY_REMOVED = "EntityRemoved"
    MESH_UPLOADED = "MeshUploaded"
    MATERIAL_UPDATED = "MaterialUpdated"
    LIGHT_UPDATED = "LightUpdated"
    CAMERA_UPDATED = "CameraUpdated"
    TEXTURE_UPLOADED = "TextureUploaded"


# ---------------------------------------------------------------------------
# Payload dataclasses (mirror Rust structs)
# ---------------------------------------------------------------------------


@dataclass
class Hello:
    version: int = PROTOCOL_VERSION
    client: str = "blender_addon"
    capabilities: list[str] = field(default_factory=lambda: ["ping", "scene_sync"])


@dataclass
class HelloAck:
    version: int = PROTOCOL_VERSION
    server: str = SERVER_NAME
    accepted: bool = True
    capabilities: list[str] = field(default_factory=lambda: ["ping"])
    reason: Optional[str] = None


@dataclass
class Ping:
    seq: int = 0
    ts_micros: int = 0


@dataclass
class Pong:
    seq: int = 0
    client_ts_micros: int = 0
    server_ts_micros: int = 0


@dataclass
class Error:
    code: str = "INTERNAL"
    message: str = ""


@dataclass
class Goodbye:
    reason: str = ""


@dataclass
class TransformUpdated:
    id: str = ""
    matrix: list[float] = field(default_factory=lambda: [0.0] * 16)
    color: Optional[list[float]] = None
    metallic: Optional[float] = None
    roughness: Optional[float] = None
    albedo_path: Optional[str] = None
    normal_path: Optional[str] = None
    metallic_path: Optional[str] = None
    roughness_path: Optional[str] = None
    emission_color: Optional[list[float]] = None
    emission_strength: Optional[float] = None


# ---------------------------------------------------------------------------
# Error codes (matching Rust reactor_bridge::protocol::codes)
# ---------------------------------------------------------------------------

CODES = {
    "INCOMPATIBLE_VERSION": "INCOMPATIBLE_VERSION",
    "UNKNOWN_MESSAGE": "UNKNOWN_MESSAGE",
    "MALFORMED_PAYLOAD": "MALFORMED_PAYLOAD",
    "NOT_HANDSHAKED": "NOT_HANDSHAKED",
    "INTERNAL": "INTERNAL",
}


# ---------------------------------------------------------------------------
# Serialization helpers
# ---------------------------------------------------------------------------


_PAYLOAD_MAP = {
    MessageType.HELLO: Hello,
    MessageType.HELLO_ACK: HelloAck,
    MessageType.PING: Ping,
    MessageType.PONG: Pong,
    MessageType.ERROR: Error,
    MessageType.GOODBYE: Goodbye,
    MessageType.TRANSFORM_UPDATED: TransformUpdated,
}


def make_message(msg_type: str, data: dict) -> str:
    """Build a JSON wire message: {"type": msg_type, "data": data}."""
    return json.dumps({"type": msg_type, "data": data}, ensure_ascii=False)


def parse_message(wire: str) -> tuple[str, dict]:
    """Parse a JSON wire message into (type, data_dict).

    Raises ValueError on malformed input.
    """
    try:
        obj = json.loads(wire)
    except json.JSONDecodeError as e:
        raise ValueError(f"JSON parse error: {e}") from e

    if not isinstance(obj, dict) or "type" not in obj:
        raise ValueError("Message missing 'type' field")

    msg_type = obj["type"]
    data = obj.get("data", {})
    if not isinstance(data, dict):
        raise ValueError("Message 'data' field must be a dict")

    return msg_type, data


def now_micros() -> int:
    """Current time in microseconds since UNIX_EPOCH."""
    return int(time.time() * 1_000_000)
